logical/framework/openapi.go

```diff
@@ -0,0 +1,413 @@
+								Schema: s,
+							},
+						},
+					}
+				}
+			}
+
+			// LIST is represented as GET with a `list` query parameter
+			if opType == logical.ListOperation || (opType == logical.ReadOperation && operations[logical.ListOperation] != nil) {
+				op.Parameters = append(op.Parameters, openapi.Parameter{
+					Name:        "list",
+					Description: "Return a list if `true`",
+					In:          "query",
+					Schema:      &openapi.Schema{Type: "string"},
+				})
+			}
+
+			// Add tags based on backend type
+			var tags []string
+			switch backendType {
+			case logical.TypeLogical:
+				tags = []string{"secrets"}
+			case logical.TypeCredential:
+				tags = []string{"auth"}
+			}
+
+			op.Tags = append(op.Tags, tags...)
+
+			// Set default responses.
+			if len(props.Responses) == 0 {
+				if opType == logical.DeleteOperation {
+					op.Responses["204"] = openapi.StdRespNoContent
+				} else {
+					op.Responses["200"] = openapi.StdRespOK
+				}
+			}
+
+			// Add any defined response details.
+			for code, responses := range props.Responses {
+				var description string
+				content := make(openapi.Content)
+
+				for i, resp := range responses {
+					if i == 0 {
+						description = resp.Description
+					}
+					if resp.Example != nil {
+						mediaType := resp.MediaType
+						if mediaType == "" {
+							mediaType = "application/json"
+						}
+
+						// Only one example per media type is allowed, so first one wins
+						if _, ok := content[mediaType]; !ok {
+							content[mediaType] = &openapi.MediaTypeObject{
+								Schema: &openapi.Schema{
+									Example: cleanResponse(resp.Example),
+								},
+							}
+						}
+					}
+				}
+
+				// a nil pointer when empty is needed for omitempty to work
+				var c *openapi.Content
+				if len(content) > 0 {
+					c = &content
+				}
+				op.Responses[code] = &openapi.Response{
+					Description: description,
+					Content:     c,
+				}
+			}
+
+			switch opType {
+			case logical.CreateOperation, logical.UpdateOperation:
+				pi.Post = op
+			case logical.ReadOperation, logical.ListOperation:
+				pi.Get = op
+			case logical.DeleteOperation:
+				pi.Delete = op
+			}
+		}
+
+		doc.Paths["/"+path] = &pi
+	}
+
+	return nil
+}
+
+func specialPathMatch(path string, specialPaths []string) bool {
+	// Test for exact or prefix match of special paths.
+	for _, sp := range specialPaths {
+		if sp == path ||
+			(strings.HasSuffix(sp, "*") && strings.HasPrefix(path, sp[0:len(sp)-1])) {
+			return true
+		}
+	}
+	return false
+}
+
+// expandPattern expands a regex pattern by generating permutations of any optional parameters
+// and changing named parameters into their {openapi} equivalents.
+func expandPattern(pattern string) []string {
+	var paths []string
+
+	// This construct is added by GenericNameRegex and is much easier to remove now
+	// than to compensate for in the other regexes.
+	pattern = strings.Replace(pattern, `\w(([\w-.]+)?\w)?`, "", -1)
```

Comment:
Could we add a test around this regex, as well? And also, would it be possible to more directly tie it to the `GenericNameRegex` in case it changes?

Reply:
I've reworked this to use whatever is currently implemented in `GenericNameRegex`, and I've updated a test to use that function as well.